src/raftstore/store/peer_storage.rs

```diff
@@ -325,46 +325,89 @@ impl InvokeContext {
-        try!(wb.put_msg_cf(
-            handle,
-            &keys::raft_state_key(self.region_id),
-            &self.raft_state
-        ));
+    pub fn save_raft_state_to(&self, raft_wb: &mut WriteBatch) -> Result<()> {
+        try!(raft_wb.put_msg(&keys::raft_state_key(self.region_id), &self.raft_state));
         Ok(())
     }
 
     #[inline]
-    pub fn save_apply_to(&self, db: &DB, wb: &mut WriteBatch) -> Result<()> {
-        let handle = try!(rocksdb::get_cf_handle(db, CF_RAFT));
-        try!(wb.put_msg_cf(
-            handle,
-            &keys::apply_state_key(self.region_id),
-            &self.apply_state
+    pub fn save_snapshot_raft_state_to(
+        &self,
+        snapshot_index: u64,
+        kv_wb: &mut WriteBatch,
+    ) -> Result<()> {
+        let mut snapshot_raft_state = self.raft_state.clone();
+        snapshot_raft_state
+            .mut_hard_state()
+            .set_commit(snapshot_index);
+        snapshot_raft_state.set_last_index(snapshot_index);
+        try!(kv_wb.put_msg(
+            &keys::snapshot_raft_state_key(self.region_id),
+            &snapshot_raft_state
         ));
         Ok(())
     }
+
+    #[inline]
+    pub fn save_apply_state_to(&self, kv_wb: &mut WriteBatch) -> Result<()> {
+        try!(kv_wb.put_msg(&keys::apply_state_key(self.region_id), &self.apply_state));
+        Ok(())
+    }
 }
 
-fn init_raft_state(engine: &DB, region: &Region) -> Result<RaftLocalState> {
-    Ok(match try!(
-        engine.get_msg_cf(CF_RAFT, &keys::raft_state_key(region.get_id()))
-    ) {
+pub fn recover_from_applying_state(kv_engine: &DB, raft_engine: &DB, region_id: u64) -> Result<()> {
+    let snapshot_raft_state_key = keys::snapshot_raft_state_key(region_id);
+    let snapshot_raft_state: RaftLocalState =
+        match box_try!(kv_engine.get_msg(&snapshot_raft_state_key)) {
+            Some(state) => state,
+            None => {
+                return Err(box_err!(
+                    "[region {}] failed to get raftstate from kv engine, \
+                     when recover from applying state",
+                    region_id
+                ));
+            }
+        };
+
+    let raft_state_key = keys::raft_state_key(region_id);
+    let raft_state: RaftLocalState = match box_try!(raft_engine.get_msg(&raft_state_key)) {
+        Some(state) => state,
+        None => RaftLocalState::new(),
+    };
+
+    // if we recv append log when applying snapshot, last_index in raft_local_state will
+    // larger than snapshot_index. since raft_local_state is written to raft engine, and
+    // raft write_batch is written after kv write_batch, raft_local_state may wrong if
+    // restart happen between the two write. so we copy raft_local_state to kv engine
+    // (snapshot_raft_state), and set snapshot_raft_state.last_index = snapshot_index.
+    // after restart, we need check last_index.
+    if last_index(&snapshot_raft_state) > last_index(&raft_state) {
```

Comment:
You said we need to overwrite raft_local_state when raft_local_state's last_index is larger than snapshot_index above, but here you do it when raft_local_state's last_index is smaller than snapshot_index.